cmd/metal-api/internal/service/async-actor.go

```diff
@@ -0,0 +1,167 @@
+	actor := &asyncActor{
+		Logger:       l,
+		IPAMer:       ip,
+		RethinkStore: ds,
+	}
+	var err error
+	_, actor.machineReleaser, err = ep.Function("releaseMachineNetworks", actor.releaseMachineNetworks)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create async bus function for machine releasing: %w", err)
+	}
+	_, actor.ipReleaser, err = ep.Function("releaseIP", actor.releaseIP)
+	if err != nil {
+		return nil, fmt.Errorf("cannot create bus function for ip releasing: %w", err)
+	}
+	return actor, nil
+}
+
+func (a *asyncActor) freeMachine(pub bus.Publisher, m *metal.Machine) error {
+	if m.State.Value == metal.LockedState {
+		return fmt.Errorf("machine is locked")
+	}
+
+	err := deleteVRFSwitches(a.RethinkStore, m, a.Logger)
+	if err != nil {
+		return err
+	}
+
+	err = publishDeleteEvent(pub, m, a.Logger)
+	if err != nil {
+		return err
+	}
+
+	// call the releaser async
+	err = a.machineReleaser(m)
+	if err != nil {
+		// log error, but what should we do here? we already called
+		// deleteVRFSwitches and publishDeleteEvent, so should we return
+		// an error or "fall through"?
+		a.Error("cannot call async machine cleanup", zap.Error(err))
+	}
+
+	old := *m
+
+	m.Allocation = nil
+	m.Tags = nil
+
+	err = a.UpdateMachine(&old, m)
+	if err != nil {
+		return err
+	}
+	a.Info("freed machine", zap.String("machineID", m.ID))
+
+	return nil
+}
+
+func (a *asyncActor) releaseMachineNetworks(machine *metal.Machine) error {
+	if machine.Allocation == nil {
+		return nil
+	}
+	for _, machineNetwork := range machine.Allocation.MachineNetworks {
+		for _, ipString := range machineNetwork.IPs {
+			ip, err := a.FindIPByID(ipString)
+			if err != nil {
+				return err
+			}
+			// ignore ips that were associated with the machine for allocation but the association is not present anymore at the ip
+			if !ip.HasMachineId(machine.GetID()) {
+				continue
+			}
+			// disassociate machine from ip
+			new := *ip
+			new.RemoveMachineId(machine.GetID())
+			err = a.UpdateIP(ip, &new)
+			if err != nil {
+				return err
+			}
+			// static ips should not be released automatically
+			if ip.Type == metal.Static {
+				continue
+			}
+			// ips that are associated to other machines will should not be released automatically
+			if len(new.GetMachineIds()) > 0 {
+				continue
+			}
+
+			// at this point the machine is removed from the IP, so the release of the
+			// IP must not fail, because this loop will never come to this point again because the
+			// begin of this loop checks if the IP contains the machine.
+			// so we fork a new job to delete the IP. if this fails .... well then ("houston we have a problem")
+			// we do not report the error to the caller, because this whole function cannot be re-do'ed.
+			a.Infow("async release IP", "ip", *ip)
+			if err := a.ipReleaser(*ip); err != nil {
+				// what should we do here? this error shows a problem with the nsq-bus system
```

Comment:
This is probably a case for our "garbage collector" or IPAM syncer or whatever we want to call that. An algorithm that finds such IP addresses that belong to "nothing" and cleans them up. I also do not know how to deal with this more adequately than that.